dbt/adapters/athena/session.py

```diff
@@ -11,3 +24,254 @@ def get_boto3_session(connection: Connection) -> boto3.session.Session:
+        """
+        Initialize the AthenaSparkSessionManager instance.
+
+        Args:
+            credentials (Any): The credentials to be used.
+            timeout (int): The timeout value in seconds.
+            polling_interval (float): The polling interval in seconds.
+            engine_config (Dict[str, int]): The engine configuration.
+
+        """
+        self.credentials = credentials
+        self.timeout = timeout
+        self.polling_interval = polling_interval
+        self.engine_config = engine_config
+        self.lock = threading.Lock()
+
+    @cached_property
+    def spark_threads(self) -> int:
+        """
+        Get the number of Spark threads.
+
+        Returns:
+            int: The number of Spark threads. If not found in the profile, returns the default thread count.
+        """
+        if not self.credentials.spark_threads:
+            LOGGER.debug(
+                f"""Spark threads not found in profile. Got: {self.credentials.spark_threads}.
+                Using default count: {DEFAULT_THREAD_COUNT}"""
+            )
+            return DEFAULT_THREAD_COUNT
+        return int(self.credentials.spark_threads)
+
+    @cached_property
+    def spark_work_group(self) -> str:
+        """
+        Get the Spark work group.
+
+        Returns:
+            str: The Spark work group. Raises an exception if not found in the profile.
+        """
+        if not self.credentials.spark_work_group:
+            raise DbtRuntimeError(f"Expected spark_work_group in profile. Got: {self.credentials.spark_work_group}")
+        return str(self.credentials.spark_work_group)
+
+    @cached_property
+    def athena_client(self) -> Any:
+        """
+        Get the AWS Athena client.
+
+        This function returns an AWS Athena client object that can be used to interact with the Athena service.
+        The client is created using the region name and profile name provided during object instantiation.
+
+        Returns:
+            Any: The Athena client object.
+
+        """
+        return get_boto3_session_from_credentials(self.credentials).client(
+            "athena", config=get_boto3_config(num_retries=self.credentials.effective_num_retries)
+        )
+
+    def get_new_sessions(self) -> List[UUID]:
+        """
+        Retrieves a list of new sessions by subtracting the existing sessions from the complete session list.
+        If no new sessions are found a new session is created provided the number of sessions is less than the
+        number of allowed spark threads.
+
+        Returns:
+            List[UUID]: A list of new session UUIDs.
+
+        """
+        sessions = self.list_sessions()
+        existing_sessions = set(spark_session_locks.keys())
+        new_sessions = [session for session in sessions if session not in existing_sessions]
+
+        if len(new_sessions) == 0:
+            if len(spark_session_locks) < self.spark_threads:
```

Comment:
@Avinash-1394 Wouldn't spark_session_locks have terminated sessions as well? Shouldnt we substract those terminated sessions before we can check against the thread count limit?

Reply:
You are right. When the spark_session_locks is first built it will only fetch sessions that are not terminated but we don't remove from it if for some reason the session is terminated while executing the models. I see that you have handled this in the PR to this branch. Hopefully that handles it!